### PR TITLE
fix: add global leaks detection on watch-run

### LINKED ISSUE #1948: 🐛 Bug: --check-leaks only detects leak the first time (e.g. when running twice with `--watch`)
I have the following bit of code, for which mocha correctly detects a leak:

```
describe('leak detection', function() {

    function testMe(){
        x = 123; // here I add 'var ' to make the test pass
    }

    it('should not leak', function() {
        testMe();
    });
});
```

By substituting 'x' for 'var x' the test passes. Seemingly, the same applies when running mocha with the --watch task, but when I remove the `var` statement the test still passes without any error about leaking. This is a bug.

```
$ mocha --check-leaks --watch

  leak detection
    ✓ should not leak
    1) should not leak

  1 passing (13ms)
  1 failing

  1) leak detection should not leak:
     Error: global leak detected: x

  leak detection
    ✓ should not leak

  1 passing (1ms)

  leak detection
    ✓ should not leak

  1 passing (2ms)
```

Using mocha 2.3.3 on Node 4.2.1

### PR DESCRIPTION
**Pull request by anpaul0615**
### Description of the Change

- apply global variable reset process when rerunning test in watch mode
- add some test (that detect global leaks and rerun that test through touch that test file, and then detect global leak again)

Fixes #1948.

### CODE REVIEW

**Comment by outsideris** [comment]
ref: #1948

File: test/integration/fixtures/options/watch/test-global-leak.fixture.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  }
+
+  it('should not leak', function(done) {
+    testMe();
+    done(); // just pass
+  });
+
+  before(function() {});
+
+  after(function() {});
```
Comment (outsideris) [inline test/integration/fixtures/options/watch/test-global-leak.fixture.js]: hooks needed here?

File: lib/cli/watch-run.js [hunk 2]
```diff
@@ -114,6 +121,15 @@ const createRerunner = (mocha, beforeRun) => {
   };
 
   const rerun = () => {
+    if (defaultGlobals) {
+      Object.keys(global).forEach(k => {
+        if (defaultGlobals.indexOf(k) === -1) {
```
Comment (outsideris) [inline lib/cli/watch-run.js]: use `if (~defaultGlobals.indexOf(k)) {` like https://github.com/mochajs/mocha/blob/master/lib/runner.js#L215 .
same code style is better.

File: lib/cli/watch-run.js [hunk 3]
```diff
@@ -85,10 +85,17 @@ const createRerunner = (mocha, beforeRun) => {
 
   let rerunScheduled = false;
 
+  let defaultGlobals = null;
```
Comment (outsideris) [inline lib/cli/watch-run.js]: Can we use another name? Because I feel it is not "default".

File: test/integration/fixtures/options/watch/test-global-leak.fixture.js (hunk 1 above)
Comment (anpaul0615) [reply test/integration/fixtures/options/watch/test-global-leak.fixture.js]: because `Runner.prototype.checkGlobals()` is invoked every test-end-event and hook-end-event. (https://github.com/mochajs/mocha/blob/master/lib/runner.js#L136-L141)

File: lib/cli/watch-run.js (hunk 2 above)
Comment (anpaul0615) [reply lib/cli/watch-run.js]: ok. I got it.

File: lib/cli/watch-run.js (hunk 3 above)
Comment (anpaul0615) [reply lib/cli/watch-run.js]: how about "prevGlobals" or "memoGlobals"?

**Comment by JoshuaKGoldberg** [comment]
👋 coming back to this @anpaul0615, are you still interested in working on this PR? As of https://github.com/mochajs/mocha/issues/5027 there are again maintainers who can review it now. No worries if you no longer have time - but if you do that'd be great!

_(I do see that your last message was asking for a naming suggestion - unless directed otherwise we'll try to review it once we're ramped up enough!)_

File: test/integration/fixtures/options/watch/test-global-leak.fixture.js (hunk 1 above)
Comment (JoshuaKGoldberg) [reply test/integration/fixtures/options/watch/test-global-leak.fixture.js:13]: That makes sense, but isn't intuitive from the test. Request: how about a quick comment explaining why these are here?

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great! 🙌 

Leaving open and approved for an additional look by @mochajs/maintenance-crew. But only a couple of small nitpicks from me.

File: lib/cli/watch-run.js
Comment (JoshuaKGoldberg) [inline lib/cli/watch-run.js:1]: @Uzlopak @voxpelli and I talked about this approach: we think it's promising and gets the job done partially, but there are blocking edge cases around side effects. For example, if the first time a test is run, the test imports cached polyfills that add globals, deleting the global after the test would mean subsequent tests would fail. `delete global[k]` isn't something that Mocha can reasonably do.

Instead, we think a more workable approach would be to have the runner check for global leaks against the initial state of globals _before all tests_. I.e. keep the `initialGlobals` idea, but not the `delete global[k]` idea. Doing so would mean that necessary globals that can only be introduced exactly one won't be deleted on `--watch`. _(does that make sense?)_

One drawback of this proposed approach is that if a leak exists in a test, then is fixed, it'll still be reported on re-runs. Mocha's messaging when `--check-leaks` _and_ `--watch` are enabled will have to note that if the leak is fixed then the user will have to restart Mocha.

So, two requests, please:

* fix the behavior so that globals are never deleted, and instead runner checks leaks against the same initial globals every time
* add a unit test that adding a global variable only in the first run is still available in subsequent runs

See also #4954 / https://github.com/mochajs/mocha/issues/4954#issuecomment-1942016336. Part of why we're looking at deprecating the `--check-leaks` feature long-term is that it isn't comprehensive enough to get many common modern use cases. The only reliable way for a test framework to achieve true encapsulation(-ish) and leak detection is with VMs the way Jest & co do. For Mocha, you'd have to build a plugin and/or use linting and/or / type checking and/or `"use strict";` and/or proper ESM.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks again for working on this - requesting some design changes. Please let us know if we've missed something! 🙂

**Comment by JoshuaKGoldberg** [comment]
👋 ping @anpaul0615, is this still something you have time for?

**Comment by JoshuaKGoldberg** [comment]
It's been over a month since last ping, so closing this out as stale. If you again have time for this & the issue hasn't yet been fixed, feel free to reopen @anpaul0615.

If anybody else sees this PR is closed and wants to tackle it, great! Go ahead and send your own PR. If it includes substantial code from this one, please include a co-author attribution.

Cheers! 🤎

**Comment by mark-wiemer** [comment]
@mostafaNazari702 we can build off this PR (we will give co-author attribution of course!)

**Comment by mostafaNazari702** [comment]
> @mostafaNazari702 we can build off this PR (we will give co-author attribution of course!)

Sure, Absolutely, i will get started ASAP.
